docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,630 @@
+- This will need to be acquired to access this state and at the beginning of many methods that access any of that state.
+- How to emit events on the main thread?
+- Does not address goal 2.
+    
+### Just access `data` asynchronously
+
+- Targets main cause of unresponsiveness (i.e. reading data).
+- No events are emitted on the non-main thread.
+- Less lazy when cancelling is possible (i.e. we do more work on the main thread before submitting the async task).
+- Splits up slicing logic into pre/post data reading, making program flow harder to follow.
+- Does not address goal 2.
+    
+### Use `QThread` and similar utilities instead of `concurrent.futures`
+
+- Standard way for plugins to support long running operations.
+- Can track progress and allow more opportunity for cancellation with `yielded` signal.
+- Can easily process done callback (which might update Qt widgets) on main thread.
+- Need to define our own task queue to achieve lazy slicing.
+- Need to connect a `QObject`, which ties our core to Qt, unless the code that controls threads does not live in core.
+    
+### Use `asyncio` package instead of `concurrent.futures`
+
+- May improve general readability of async code for some.
+- Mostly syntactic sugar on top of `concurrent.futures`.
+- Likely need an `asyncio` event loop distinct from Qt's main event loop, which could be confusing and cause issues.
+
+
+## Discussion
+
+- [Initial announcement on Zulip](https://napari.zulipchat.com/#narrow/stream/296574-working-group-architecture/topic/Async.20slicing.20project).
+    - Consider (re)sampling instead of slicing as the name for the operation discussed here.  
+- [Problems with `NAPARI_ASYNC=1`](https://forum.image.sc/t/even-with-napari-async-1-data-loading-is-blocking-the-ui-thread/68097/4)
+    - The existing experimental async code doesn't handle some seemingly simple usage.
+- [Remove slice state from layer](https://github.com/napari/napari/issues/4682)
+    - Feature request to remove slice/render state from the layer types.
+    - Motivated by creating multiple slices of the same layers for multiple canvases.
+    
+### Open questions
+
+- Should `Dims.current_step` represent the last slice position request or the last slice response?
+    - With sync slicing, there is no distinction.
+    - If it represents the last slice response, then what should we connect the sliders to?
+    - Similar question for `corner_pixels` and others.
+
+- Which new classes, methods, and attributes should be private?
+    - Maybe all of them in the short term, so that we can wait for things to mature before committing to a public API.
+
+- Does this design work well with slicing operations other than slider movements?
+    - We mostly use moving the slider as a driving example here, but there are other things that cause slicing.
+    - E.g. changing the display dimensionality, panning, zooming, explicit call to `refresh`.
```

Comment:
I think there's a big difference between if `Dims` changes, which is less frequent/ discontinuous vs if `Camera` changes which is continuous/ you want smooth. If hooking up to camera we probably either need some really fast synchronous code to determine if there have been large enough FOV changes to trigger re-slicing, or get to async as quickly as possible and then not make slicing request if FOV changes are small. Unclear if this can be done in a global way or if it needs to be layer specific (say aware of a layers scales). As most of these considerations are for multiscale fine to ignore now I think, but just wanted to mention

Reply:
I wonder how much we could get away with just requesting a little more data than is visible in the field of view so that it's hard to pan out of what is currently visible. Not super efficient, but easy to implement!

Reply:
That's how the current multiscale works, but I think the approach in the experimental octree implementation is more what we want longer term, where we have a fixed sense of the tiles that space you render in (we called them chunks in the experimental octree implementation) and then you pull in the chunks you need to cover the viewed space. This is a little more tricky for datasets where we don't start with level-of-detail tree that exactly matches what we assume, but I think still would be the way to go